src/cryptonote_basic/hardfork.cpp

```diff
@@ -27,468 +26,136 @@
-    accumulated_votes += last_versions[v];
-    uint32_t threshold = (window_size * heights[n].threshold + 99) / 100;
-    if (height >= heights[n].height && accumulated_votes >= threshold) {
-      return n;
+    } else if (it->version == version) {
+      found.first = it->height;
     }
   }
-  return current_fork_index;
-}
-
-HardFork::State HardFork::get_state(time_t t) const
-{
-  std::unique_lock l{lock};
-
-  // no hard forks setup yet
-  if (heights.size() <= 1)
-    return Ready;
-
-  time_t t_last_fork = heights.back().time;
-  if (t >= t_last_fork + forked_time)
-    return LikelyForked;
-  if (t >= t_last_fork + update_time)
-    return UpdateNeeded;
-  return Ready;
-}
-
-HardFork::State HardFork::get_state() const
-{
-  return get_state(time(NULL));
-}
-
-uint8_t HardFork::get(uint64_t height) const
-{
-  std::unique_lock l{lock};
-  if (height > db.height()) {
-    assert(false);
-    return INVALID_HF_VERSION;
-  }
-  if (height == db.height()) {
-    return get_current_version();
-  }
-  return db.get_hard_fork_version(height);
+  return found;
 }
 
-uint8_t HardFork::get_current_version() const
-{
-  std::unique_lock l{lock};
-  return heights[current_fork_index].version;
-}
-
-uint8_t HardFork::get_ideal_version() const
-{
-  std::unique_lock l{lock};
-  return heights.back().version;
-}
-
-uint8_t HardFork::get_ideal_version(uint64_t height) const
-{
-  std::unique_lock l{lock};
-  for (unsigned int n = heights.size() - 1; n > 0; --n) {
-    if (height >= heights[n].height) {
-      return heights[n].version;
-    }
+std::pair<uint8_t, uint8_t>
+get_network_version_revision(network_type nettype, uint64_t height) {
+  std::pair<uint8_t, uint8_t> result;
+  for (auto [it, end] = get_hard_forks(nettype); it != end; it++) {
+    if (it->height <= height)
+      result = {it->version, it->snode_revision};
+    else
+      break;
   }
-  return original_version;
+  return result;
 }
 
-uint64_t HardFork::get_earliest_ideal_height_for_version(uint8_t version) const
-{
-  uint64_t height = std::numeric_limits<uint64_t>::max();
-  for (auto i = heights.rbegin(); i != heights.rend(); ++i) {
-    if (i->version >= version) {
-      height = i->height;
-    } else {
-      break;
-    }
-  }
-  return height;
+bool is_hard_fork_at_least(network_type type, uint8_t version, uint64_t height) {
+  return get_network_version(type, height) >= version;
 }
 
-uint8_t HardFork::get_next_version() const
+std::pair<uint8_t, uint8_t>
+get_ideal_block_version(network_type nettype, uint64_t height)
 {
-  std::unique_lock l{lock};
-  uint64_t height = db.height();
-  for (auto i = heights.rbegin(); i != heights.rend(); ++i) {
-    if (height >= i->height) {
-      return (i == heights.rbegin() ? i : (i - 1))->version;
-    }
+  std::pair<uint8_t, uint8_t> result;
+  for (auto [it, end] = get_hard_forks(nettype); it != end; it++) {
+    if (it->height <= height)
+      result.first = it->height;
+    result.second = it->height;
```

Comment:
`result.second = it->snode_revision;`

Reply:
Actually it has to be `it->height` because current nodes won't allow a minor version < the major version :(